File: common/src/sram23lc1024.c
#include "sram23lc1024.h"
#include "ssp.h"
#include "systick.h"

void sram_init(void)
{
    ssp0_init();
}

uint8_t sram_read_mode(void)
{
    uint8_t v;
    ssp0_cs_low();
    ssp0_xfer(0x05u);
    v = ssp0_xfer(0xFFu);
    ssp0_cs_high();
    return v;
}

void sram_write_mode(uint8_t v)
{
    ssp0_cs_low();
    ssp0_xfer(0x01u);
    ssp0_xfer(v);
    ssp0_cs_high();
}

int sram_test_simple(void)
{
    static uint8_t wbuf[256];
    static uint8_t rbuf[256];
    const uint8_t patterns[] = { 0xAAu, 0x55u };
    const uint32_t sram_size = 128u * 1024u;
    const uint32_t chunk = sizeof(wbuf);

    for (unsigned int p = 0; p < sizeof(patterns); p++) {
        uint8_t pat = patterns[p];

        ssp0_cs_low();
        ssp0_xfer(0x02u);
        ssp0_xfer(0x00u);
        ssp0_xfer(0x00u);
        ssp0_xfer(0x00u);
        for (uint32_t addr = 0; addr < sram_size; addr += chunk) {
            for (uint32_t i = 0; i < chunk; i++) {
                wbuf[i] = pat;
                ssp0_xfer(wbuf[i]);
            }
        }
        ssp0_cs_high();

        ssp0_cs_low();
        ssp0_xfer(0x03u);
        ssp0_xfer(0x00u);
        ssp0_xfer(0x00u);
        ssp0_xfer(0x00u);
        for (uint32_t addr = 0; addr < sram_size; addr += chunk) {
            for (uint32_t i = 0; i < chunk; i++) {
                rbuf[i] = ssp0_xfer(0xFFu);
            }
            for (uint32_t i = 0; i < chunk; i++) {
                if (rbuf[i] != pat) {
                    ssp0_cs_high();
                    return -1;
                }
            }
        }
        ssp0_cs_high();
    }

    return 0;
}

void sram_bandwidth_test(uint32_t *write_kb_s, uint32_t *write_ms,
                         uint32_t *read_kb_s, uint32_t *read_ms)
{
    const uint32_t sram_size = 128u * 1024u;
    const uint32_t chunk = 256u;
    uint32_t start_ms;
    uint32_t elapsed_ms;
    uint32_t bytes_per_s;

    /* Write bandwidth */
    start_ms = systick_millis();
    ssp0_cs_low();
    ssp0_xfer(0x02u);
    ssp0_xfer(0x00u);
    ssp0_xfer(0x00u);
    ssp0_xfer(0x00u);
    for (uint32_t addr = 0; addr < sram_size; addr += chunk) {
        for (uint32_t i = 0; i < chunk; i++) {
            ssp0_xfer(0x00u);
        }
    }
    ssp0_cs_high();
    elapsed_ms = systick_millis() - start_ms;
    if (elapsed_ms == 0u) {
        elapsed_ms = 1u;
    }
    bytes_per_s = (sram_size * 1000u) / elapsed_ms;
    *write_kb_s = bytes_per_s / 1024u;
    *write_ms = elapsed_ms;

    /* Read bandwidth */
    start_ms = systick_millis();
    ssp0_cs_low();
    ssp0_xfer(0x03u);
    ssp0_xfer(0x00u);
    ssp0_xfer(0x00u);
    ssp0_xfer(0x00u);
    for (uint32_t addr = 0; addr < sram_size; addr += chunk) {
        for (uint32_t i = 0; i < chunk; i++) {
            (void)ssp0_xfer(0xFFu);
        }
    }
    ssp0_cs_high();
    elapsed_ms = systick_millis() - start_ms;
    if (elapsed_ms == 0u) {
        elapsed_ms = 1u;
    }
    bytes_per_s = (sram_size * 1000u) / elapsed_ms;
    *read_kb_s = bytes_per_s / 1024u;
    *read_ms = elapsed_ms;
}

File: common/rust/src/sram23lc1024.rs
#![allow(dead_code)]

use crate::systick;
use crate::ssp;

pub fn init() {
    ssp::init();
}

pub fn read_mode() -> u8 {
    let v;
    ssp::cs_low();
    ssp::xfer(0x05);
    v = ssp::xfer(0xFF);
    ssp::cs_high();
    v
}

pub fn write_mode(v: u8) {
    ssp::cs_low();
    ssp::xfer(0x01);
    ssp::xfer(v);
    ssp::cs_high();
}

pub fn test_simple() -> bool {
    const SRAM_SIZE: u32 = 128 * 1024;
    const CHUNK: usize = 256;
    let patterns = [0xAAu8, 0x55u8];
    let mut wbuf = [0u8; CHUNK];
    let mut rbuf = [0u8; CHUNK];

    for pat in patterns.iter() {
        ssp::cs_low();
        ssp::xfer(0x02);
        ssp::xfer(0x00);
        ssp::xfer(0x00);
        ssp::xfer(0x00);
        let mut addr = 0u32;
        while addr < SRAM_SIZE {
            for i in 0..CHUNK {
                wbuf[i] = *pat;
                ssp::xfer(wbuf[i]);
            }
            addr += CHUNK as u32;
        }
        ssp::cs_high();

        ssp::cs_low();
        ssp::xfer(0x03);
        ssp::xfer(0x00);
        ssp::xfer(0x00);
        ssp::xfer(0x00);
        addr = 0u32;
        while addr < SRAM_SIZE {
            for i in 0..CHUNK {
                rbuf[i] = ssp::xfer(0xFF);
            }
            for i in 0..CHUNK {
                if rbuf[i] != *pat {
                    ssp::cs_high();
                    return false;
                }
            }
            addr += CHUNK as u32;
        }
        ssp::cs_high();
    }

    true
}

pub fn bandwidth_test() -> (u32, u32, u32, u32) {
    const SRAM_SIZE: u32 = 128 * 1024;
    const CHUNK: u32 = 256;
    let mut start = systick::millis();
    let mut elapsed;

    // Write bandwidth
    ssp::cs_low();
    ssp::xfer(0x02);
    ssp::xfer(0x00);
    ssp::xfer(0x00);
    ssp::xfer(0x00);
    let mut addr = 0u32;
    while addr < SRAM_SIZE {
        let mut i = 0u32;
        while i < CHUNK {
            ssp::xfer(0x00);
            i += 1;
        }
        addr += CHUNK;
    }
    ssp::cs_high();
    elapsed = systick::millis().wrapping_sub(start);
    if elapsed == 0 { elapsed = 1; }
    let write_kb_s = (SRAM_SIZE * 1000 / elapsed) / 1024;

    // Read bandwidth
    start = systick::millis();
    ssp::cs_low();
    ssp::xfer(0x03);
    ssp::xfer(0x00);
    ssp::xfer(0x00);
    ssp::xfer(0x00);
    addr = 0;
    while addr < SRAM_SIZE {
        let mut i = 0u32;
        while i < CHUNK {
            let _ = ssp::xfer(0xFF);
            i += 1;
        }
        addr += CHUNK;
    }
    ssp::cs_high();
    elapsed = systick::millis().wrapping_sub(start);
    if elapsed == 0 { elapsed = 1; }
    let read_kb_s = (SRAM_SIZE * 1000 / elapsed) / 1024;

    (write_kb_s, 0, read_kb_s, 0)
}
